File: server/apps/products/views/flash_sale.py
"""
API views for Flash Sale system.
"""
from rest_framework import generics, status
from rest_framework.views import APIView
from rest_framework.response import Response
from rest_framework.permissions import AllowAny
from django.utils import timezone
from django.db.models import F, Count
from drf_yasg.utils import swagger_auto_schema
from drf_yasg import openapi

from products.models import FlashSaleSession, FlashSaleItem
from products.serializers.flash_sale import (
    FlashSaleSessionListSerializer,
    FlashSaleSessionDetailSerializer,
    FlashSaleItemSerializer,
    CurrentFlashSaleSerializer,
)


class GetCurrentFlashSaleView(APIView):
    """
    Get the current active or upcoming flash sale.
    
    Returns:
    - current_session: Currently active flash sale (if any)
    - upcoming_session: Next scheduled flash sale (if no current)
    - featured_items: Top items from the current/upcoming session
    - server_time: Current server time for countdown sync
    """
    permission_classes = [AllowAny]

    @swagger_auto_schema(
        operation_summary="Get current flash sale",
        operation_description="Returns current or upcoming flash sale with featured items",
        responses={200: CurrentFlashSaleSerializer()}
    )
    def get(self, request):
        now = timezone.now()
        
        # Get current active session
        current_session = FlashSaleSession.objects.filter(
            status=FlashSaleSession.Status.ACTIVE,
            is_active=True,
            start_time__lte=now,
            end_time__gte=now,
        ).prefetch_related(
            'items',
            'items__product',
            'items__product__category',
            'items__product__manufacturer',
            'items__product__images',
        ).first()
        
        upcoming_session = None
        
        # If no current session, get upcoming
        if not current_session:
            upcoming_session = FlashSaleSession.objects.filter(
                status=FlashSaleSession.Status.SCHEDULED,
                is_active=True,
                start_time__gt=now,
            ).prefetch_related(
                'items',
                'items__product',
                'items__product__category',
                'items__product__manufacturer',
                'items__product__images',
            ).order_by('start_time').first()
        
        # Get featured items from current or upcoming session
        featured_items = []
        active_session = current_session or upcoming_session
        
        if active_session:
            featured_items = active_session.items.filter(
                is_active=True
            ).select_related(
                'product',
                'product__category',
                'product__manufacturer',
            ).prefetch_related(
                'product__images',
            ).order_by('sort_order', '-sold_quantity')[:8]
        
        # Build response
        response_data = {
            'current_session': current_session,
            'upcoming_session': upcoming_session,
            'featured_items': featured_items,
            'server_time': now,
        }
        
        serializer = CurrentFlashSaleSerializer(response_data)
        return Response(serializer.data)


class FlashSaleSessionListView(generics.ListAPIView):
    """
    List all flash sale sessions (current, upcoming, and recent ended).
    """
    permission_classes = [AllowAny]
    serializer_class = FlashSaleSessionListSerializer

    def get_queryset(self):
        now = timezone.now()
        status_filter = self.request.query_params.get('status', None)
        
        queryset = FlashSaleSession.objects.filter(
            is_active=True
        ).annotate(total_items_count=Count('items', distinct=True)).prefetch_related('items')
        
        if status_filter:
            queryset = queryset.filter(status=status_filter)
        else:
            # Default: show active and upcoming, plus recently ended (last 7 days)
            from datetime import timedelta
            week_ago = now - timedelta(days=7)
            queryset = queryset.filter(
                end_time__gte=week_ago
            )
        
        return queryset.order_by('-start_time')


class FlashSaleSessionDetailView(generics.RetrieveAPIView):
    """
    Get flash sale session details by slug.
    Includes all items in the session.
    """
    permission_classes = [AllowAny]
    serializer_class = FlashSaleSessionDetailSerializer
    lookup_field = 'slug'

    def get_queryset(self):
        return FlashSaleSession.objects.filter(
            is_active=True
        ).prefetch_related(
            'items',
            'items__product',
            'items__product__category',
            'items__product__manufacturer',
            'items__product__images',
        )


class FlashSaleItemListView(generics.ListAPIView):
    """
    List all items in a specific flash sale session.
    Supports filtering by availability.
    """
    permission_classes = [AllowAny]
    serializer_class = FlashSaleItemSerializer

    @swagger_auto_schema(
        manual_parameters=[
            openapi.Parameter('session', openapi.IN_QUERY, description="Session slug", type=openapi.TYPE_STRING, required=True),
            openapi.Parameter('available_only', openapi.IN_QUERY, description="Show only available items", type=openapi.TYPE_BOOLEAN),
        ]
    )
    def get(self, request, *args, **kwargs):
        return super().get(request, *args, **kwargs)

    def get_queryset(self):
        session_slug = self.request.query_params.get('session')
        available_only = self.request.query_params.get('available_only', '').lower() in ('true', '1')
        
        if not session_slug:
            return FlashSaleItem.objects.none()
        
        queryset = FlashSaleItem.objects.filter(
            session__slug=session_slug,
            session__is_active=True,
            is_active=True,
        ).select_related(
            'product',
            'product__category',
            'product__manufacturer',
            'session',
        ).prefetch_related(
            'product__images',
        ).order_by('sort_order', '-sold_quantity')
        
        if available_only:
            queryset = queryset.filter(remaining_quantity__gt=0)
        
        return queryset


class FlashSaleItemDetailView(generics.RetrieveAPIView):
    """
    Get a specific flash sale item by ID.
    """
    permission_classes = [AllowAny]
    serializer_class = FlashSaleItemSerializer
    lookup_field = 'pk'

    def get_queryset(self):
        return FlashSaleItem.objects.filter(
            is_active=True,
            session__is_active=True,
        ).select_related(
            'product',
            'product__category',
            'product__manufacturer',
            'session',
        ).prefetch_related(
            'product__images',
        )


class CheckFlashSaleForProductView(APIView):
    """
    Check if a product has an active flash sale.
    Used by product detail page to show flash sale price.
    """
    permission_classes = [AllowAny]

    @swagger_auto_schema(
        manual_parameters=[
            openapi.Parameter('product_id', openapi.IN_QUERY, description="Product ID (UUID)", type=openapi.TYPE_STRING, required=True),
        ],
        responses={
            200: openapi.Response(
                description="Flash sale info if available",
                examples={
                    "application/json": {
                        "has_flash_sale": True,
                        "flash_sale_item": {
                            "id": "...",
                            "flash_sale_price": 50000,
                            "original_price": 100000,
                            "discount_percentage": 50,
                            "remaining_quantity": 10,
                            "session_end_time": "2024-01-15T23:59:59Z"
                        }
                    }
                }
            )
        }
    )
    def get(self, request):
        product_id = request.query_params.get('product_id')
        
        if not product_id:
            return Response(
                {'error': 'product_id is required'},
                status=status.HTTP_400_BAD_REQUEST
            )
        
        now = timezone.now()
        
        # Find active flash sale item for this product
        flash_sale_item = FlashSaleItem.objects.filter(
            product_id=product_id,
            is_active=True,
            remaining_quantity__gt=0,
            session__status=FlashSaleSession.Status.ACTIVE,
            session__is_active=True,
            session__start_time__lte=now,
            session__end_time__gte=now,
        ).select_related('session').first()
        
        if flash_sale_item:
            return Response({
                'has_flash_sale': True,
                'flash_sale_item': {
                    'id': str(flash_sale_item.id),
                    'flash_sale_price': float(flash_sale_item.flash_sale_price),
                    'original_price': float(flash_sale_item.original_price),
                    'discount_percentage': flash_sale_item.discount_percentage,
                    'remaining_quantity': flash_sale_item.remaining_quantity,
                    'max_per_user': flash_sale_item.max_per_user,
                    'sold_percentage': flash_sale_item.sold_percentage,
                    'session_name': flash_sale_item.session.name,
                    'session_end_time': flash_sale_item.session.end_time.isoformat(),
                    'time_remaining': flash_sale_item.session.time_remaining,
                }
            })
        
        return Response({
            'has_flash_sale': False,
            'flash_sale_item': None
        })

File: mcp/banthuoc-api-server/src/apis/flash-sale.ts
/**
 * Flash Sale API Endpoints
 *
 * Source: server/apps/products/views/flash_sale.py, flash_sale_admin.py
 *         server/apps/products/serializers/flash_sale.py
 */
import type { ApiEndpoint } from "../openapi.js";

export const endpoints: ApiEndpoint[] = [
  {
    path: "/api/flash-sale/",
    method: "GET",
    summary: "Get current active flash sale with featured items (max 8), plus upcoming session info",
    tags: ["Flash Sale"],
    auth: ["Public"],
    params: [],
    response: {
      current_session: {
        id: "uuid", name: "Flash Sale 11/11", slug: "flash-sale-11-11",
        start_time: "2026-07-08T00:00:00Z", end_time: "2026-07-09T00:00:00Z",
        status: "ACTIVE", description: "...",
        total_items_count: 25,
      },
      upcoming_session: { id: "uuid", name: "Flash Sale 12/12", status: "SCHEDULED" },
      featured_items: [{
        id: "uuid",
        product: { id: "uuid", name: "Paracetamol", price: "50000", image_url: "https://..." },
        original_price: "50000.00", flash_sale_price: "35000.00",
        discount_percentage: 30, total_quantity: 100, remaining_quantity: 45,
        sold_quantity: 55, sold_percentage: 55, max_per_user: 3,
        is_sold_out: false, is_active: true,
      }],
      server_time: "2026-07-08T12:00:00Z",
    },
  },
  {
    path: "/api/flash-sale/sessions/",
    method: "GET",
    summary: "List all flash sale sessions (active, upcoming, recently ended within 7 days)",
    tags: ["Flash Sale"],
    auth: ["Public"],
    params: [
      { name: "status", in: "query", type: "string", required: false, description: "Filter: SCHEDULED, ACTIVE, ENDED" },
    ],
    response: {
      count: 5,
      results: [{
        id: "uuid", name: "Flash Sale 11/11", slug: "flash-sale-11-11",
        status: "ACTIVE", start_time: "...", end_time: "...", total_items_count: 25,
      }],
    },
  },
  {
    path: "/api/flash-sale/sessions/{slug}/",
    method: "GET",
    summary: "Get flash sale session detail with all items",
    tags: ["Flash Sale"],
    auth: ["Public"],
    params: [{ name: "slug", in: "path", type: "string", required: true, description: "Session slug" }],
    response: {
      id: "uuid", name: "...", status: "ACTIVE",
      items: [{ id: "uuid", product: {}, flash_sale_price: "35000", discount_percentage: 30 }],
    },
  },
  {
    path: "/api/flash-sale/items/",
    method: "GET",
    summary: "List all flash sale items across sessions",
    tags: ["Flash Sale"],
    auth: ["Public"],
    params: [],
    response: [{ id: "uuid", product: {}, flash_sale_price: "35000", remaining_quantity: 45 }],
  },
  {
    path: "/api/flash-sale/items/{id}/",
    method: "GET",
    summary: "Get single flash sale item with session_name, session_end_time",
    tags: ["Flash Sale"],
    auth: ["Public"],
    params: [{ name: "pk", in: "path", type: "uuid", required: true, description: "Flash sale item ID" }],
    response: {
      id: "uuid", product: {}, original_price: "50000", flash_sale_price: "35000",
      discount_percentage: 30, remaining_quantity: 45,
      session_name: "Flash Sale 11/11", session_end_time: "2026-07-09T00:00:00Z",
    },
  },
  {
    path: "/api/flash-sale/check/",
    method: "GET",
    summary: "Check if a product is in any active flash sale",
    tags: ["Flash Sale"],
    auth: ["Public"],
    params: [{ name: "product_id", in: "query", type: "string", required: true, description: "Product ID (UUID)" }],
    response: { in_flash_sale: true, flash_sale_price: "35000", remaining: 45 },
  },
];
